### PR TITLE
Add support for double, dashed, and dotted underlines 

### LINKED ISSUE #17362: tui: add support for  drawing dashed/dotted underlines
### Feature already in Vim?

_No response_

### Feature description

Since neovim is already wired to draw undercurls, it should be possible to add dashed/dotted underlines to cover the reset of the lines mentioned in https://sw.kovidgoyal.net/kitty/underlines/ , since some terminals do support them and I don't see a reason not to?

### PR DESCRIPTION
**Pull request by kchibisov**
This commit finishes support for colored and styled underlines adding
`CSI 4 : [2,4,5] m` support.

Fixes https://github.com/neovim/neovim/issues/17362.

### CODE REVIEW

File: runtime/syntax/vim.vim [hunk 1]
```diff
@@ -574,7 +574,7 @@ syn match	vimHiBang	contained	"!"	skipwhite nextgroup=@vimHighlightCluster
 
 syn match	vimHiGroup	contained	"\i\+"
 syn case ignore
-syn keyword	vimHiAttrib	contained	none bold inverse italic nocombine reverse standout strikethrough underline undercurl
+syn keyword	vimHiAttrib	contained	none bold inverse italic nocombine reverse standout strikethrough underline undercurl lll
```
Comment (jamessan) [inline runtime/syntax/vim.vim:577]: What's this change for?

**Comment by gpanders** [comment]
Thanks for the PR!

I have a nit on the naming though. I am not sure I like `underline_double`, `underline_dotted`, etc. They are the only names in all of the attrs that use underscores, and they are quite long.

What about instead `underdot` and `underdash` for the dotted and dashed lines? This is consistent with the already existing `undercurl`. For double underline, maybe `underlineline` or `underlinedouble`?

**Comment by lewis6991** [comment]
What about `underdoubleline`?

File: runtime/syntax/vim.vim (hunk 1 above)
Comment (kchibisov) [reply runtime/syntax/vim.vim]: I left a mark for myself initially to update later on. Should be updated now.

**Comment by kchibisov** [comment]
@gpanders I've renamed to `underdot`, `underdash`, and `underlineline`.

--

Due to shortening test seems to pass now :P

File: runtime/doc/syntax.txt [hunk 2]
```diff
@@ -4894,8 +4898,9 @@ cterm={attr-list}			*attr-list* *highlight-cterm* *E418*
 
 	Note that "bold" can be used here and by using a bold font.  They
 	have the same effect.
-	"undercurl" falls back to "underline" in a terminal that does not
-	support it.  The color is set using |highlight-guisp|.
+	"underlineline", "undercurl", "underdot", and "underdash" falls back
+	to "underline" in a terminal that does not support it. The color is set
```
Comment (gpanders) [inline runtime/doc/syntax.txt:4902]:
```suggestion
	"underlineline", "undercurl", "underdot", and "underdash" fall back
	to "underline" in a terminal that does not support them. The color is set
```

File: runtime/doc/ui.txt [hunk 3]
```diff
@@ -444,14 +449,17 @@ is not active. New UIs should implement |ui-linegrid| instead.
...
 	`reverse`:	reverse video. Foreground and background colors are
 			switched.
 	`italic`:	italic text.
 	`bold`:		bold text.
 	`strikethrough`:  struckthrough text.
 	`underline`:	underlined text. The line has `special` color.
+	`underlineline`:  double underlined text. The lines has `special` color.
 	`undercurl`:	undercurled text. The curl has `special` color.
+	`underdot`:	underdotted text. The dots has `special` color.
+	`underdash`:	underdashed text. The dashes has `special` color.
```
Comment (gpanders) [inline runtime/doc/ui.txt:462]:
```suggestion
	`underlineline`:  double underlined text. The lines have `special` color.
	`undercurl`:	undercurled text. The curl has `special` color.
	`underdot`:	underdotted text. The dots have `special` color.
	`underdash`:	underdashed text. The dashes have `special` color.
```

**Review by gpanders — APPROVED** [review]
I only have some doc changes. The code LGTM, though I'm not the expert in the TUI stuff.

**Comment by lewis6991** [comment]
#17057 (overline support) relates closely to this. Can we cherry-pick the commit in that PR to here?

**Comment by kchibisov** [comment]
>https://github.com/neovim/neovim/pull/17057 (overline support) relates closely to this. Can we cherry-pick the commit in that PR to here?

I'm not sure how those relates, given that #17057 is completely different escape sequence, however this commit just finishes the support for already existing escape inside neovim, it just neovim initially decided to partially implement it. The coloring is also shouldn't be applied to overline, since it's for underlines, and overline isn't part of kitty extension.

**Comment by lewis6991** [comment]
Other than the line 1957 in tui.c for adding the escape code, everything else applies.

**Comment by kchibisov** [comment]
>But you know more than me if you feel it is out of scope.

I'm speaking from a terminal writer perspective. It's the first time I heard about overline, I'll add it inside (let's say alacritty), since it's a niche escape sequence.

The kitty extension on other side is widely adopt and mostly all modern(not even modern to this point) terminal emulators do support it.

But I'm not a neovim developer, so can't speak up more, nor that my opinion really matters.

**Comment by lewis6991** [comment]
Apparently overline is supported by Konsole and vte which seems like good enough adoption to me.

File: runtime/syntax/vim.vim [hunk 4]
```diff
@@ -574,7 +574,7 @@ syn match	vimHiBang	contained	"!"	skipwhite nextgroup=@vimHighlightCluster
 
 syn match	vimHiGroup	contained	"\i\+"
 syn case ignore
-syn keyword	vimHiAttrib	contained	none bold inverse italic nocombine reverse standout strikethrough underline undercurl
+syn keyword	vimHiAttrib	contained	none bold inverse italic nocombine reverse standout strikethrough underline underlineline undercurl underdot underdash
```
Comment (zeertzjq) [inline runtime/syntax/vim.vim:577]: trailing space

Comment (kchibisov) [reply runtime/syntax/vim.vim]: I've removed other trailing spaces in this particular file if you're fine with it.

**Review by jamessan — CHANGES_REQUESTED** [review]
Couple fixes and then it looks good to merge

**Review by jamessan — APPROVED** [review]

File: src/nvim/eval/funcs.c [hunk 5]
```diff
@@ -11409,14 +11409,22 @@ static void f_synIDattr(typval_T *argvars, typval_T *rettv, FunPtr fptr)
...
     }
     break;
-  case 'u':
-    if (STRLEN(what) <= 5 || TOLOWER_ASC(what[5]) != 'c') {  // underline
-      p = highlight_has_attr(id, HL_UNDERLINE, modec);
-    } else {  // undercurl
+  case 'u': {
+    const size_t len = STRLEN(what);
+    if (len <= 5 || (TOLOWER_ASC(what[5]) == 'l' && len <= 9)) {  // underline
       p = highlight_has_attr(id, HL_UNDERCURL, modec);
```
Comment (jamessan) [inline src/nvim/eval/funcs.c:11415]: This should be using `HL_UNDERLINE`

**Pull request merged**